eng/reproduce-vmr-sync.sh

```diff
@@ -0,0 +1,184 @@
+#!/bin/bash
+
+### This script helps to reproduce potential failures of the 'Synchronize dotnet/dotnet'
+### build step from the **VMR Source-Build** job locally.
+### The following scenario is assumed:
+### - There is a PR in dotnet/installer
+### - The PR is failing on the 'VMR Source-Build' job in the 'Synchronize dotnet/dotnet' step
+###
+### There are a few possible reasons but usually this happens because there is a mismatch between
+### source patches that are applied on top of the files that are being synchronized from repositories
+### into the 'dotnet/dotnet' repo and new changes in the repositories.
+### This manifests as the following error:
+###     fail: Failed to synchronize repo installer
+###           Failed to apply the patch for src/aspnetcore
+###           Exit code: 1
+###           Std err:
+###           error: patch failed: src/aspnetcore/eng/SourceBuild.props:55
+###           error: src/aspnetcore/eng/SourceBuild.props: patch does not apply
+###
+### where 'src/aspnetcore/eng/SourceBuild.props' would be the file that is being patched and new
+### changes to it are conflicting with the patch that is being applied.
+###
+### The whole process can be reproduced locally easily by running this script.
+### The patches are located in a folder in the 'dotnet/installer' repository.
+### At the moment of writing, the location is 'src/SourceBuild/tarball/patches' but to get
+### the up-to-date location, please see the 'patchesPath' property in
+### https://github.com/dotnet/dotnet/blob/main/src/source-mappings.json
+###
+### You will need to compare what is in the patch and what is in the file and fix the patch.
+###
+### The tooling that synchronizes the VMR will need to clone the various repositories.
+### It clones them into a temporary folder and re-uses them on future runs so it is advised
+### you dedicate a folder to this to speed up your re-runs.
+###
+### This script will synchronize the 'dotnet/dotnet' repo locally and let you inspect the changes.
+###
+### USAGE:
+###     ./reproduce-vmr-sync.sh --installer-dir "$HOME/repos/installer" --tmp-dir "$HOME/repos/tmp"
+### Options:
+###   --installer-dir
+###       Path to the 'dotnet/installer' repo which is checked out at the PR commit
+###   --tmp-dir
+###       Path to the temporary folder where the repositories will be cloned
+###   --vmr-branch
```

Comment:
We need to match that branch in the VMR and if you're on a PR branch this might break?

Reply:
Sorry - I meant the vmr-dir if specified.

Reply:
I agree with what you propose but there is a (temporary) reason I chose to do it this for now. The synchronization of the VMR creates a separate branch to which it commits intermediary progress (each repo has a commit, VMR patch reapplication get one at the end) so that the resulting git graph is something like this:

![image](https://user-images.githubusercontent.com/7013027/206245900-26a05fbe-adce-4d8a-8233-9eddd16f046c.png)

Now, I am a bit unclear on how we want this to behave for when things break halfway-through. If you break and you would re-run the script, the sync would say "installer is already at SHA x, done". So that's why I put there this "checkout main".

So what you propose would work for the first run of the script when we start off somewhere nice (release branch, main..) but not sure what then?

Reply:
We could checkout the original branch and drop whatever progress but then you cannot investigate the VMR

Reply:
hmm, this is more complicated than I thought.  We should sync on this but this discussion point should not hold up the PR.  I think the current behavior is valuable as proposed.